plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/getbean/DetachableBeanHolder.java

```diff
@@ -22,73 +22,82 @@
 import java.lang.ref.WeakReference;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
+
 import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 import org.hotswap.agent.logging.AgentLogger;
+import org.hotswap.agent.util.spring.util.CollectionUtils;
+import org.hotswap.agent.util.spring.util.StringUtils;
 
 /**
- *
  * Loadable detachable Spring bean holder
  *
  * @author Erki Ehtla
- *
  */
 public class DetachableBeanHolder implements Serializable {
 
     private static final long serialVersionUID = -7443802320153815102L;
 
+    private String beanName;
     private Object bean;
     private Object beanFactory;
-    private Class<?>[] paramClasses;
-    private Object[] paramValues;
-    private static List<WeakReference<DetachableBeanHolder>> beanProxies =
-            Collections.synchronizedList(new ArrayList<WeakReference<DetachableBeanHolder>>());
+    /**
+     * Q:Why beanProxies is a map?
+     * A:beanProxies used to be a list structure, without beanName, so it could not be cleared
+     * based on beanName, and could only be cleared in full, which caused those beans
+     * that had not changed to be cleared, which was unnecessary.
+     * -
+     * Q:Why value is a list?
+     * A:If want to let HotSwap to take effect on multiple prototype beans of the same
+     * class, a structure with one name corresponding to multiple beans is required.
+     * SpringPluginTest#hotswapPrototypeTestExistingInstance can test prototype beans.
+     */
+    private static final Map<String, List<WeakReference<DetachableBeanHolder>>> beanProxies = new ConcurrentHashMap<>();
     private static AgentLogger LOGGER = AgentLogger.getLogger(DetachableBeanHolder.class);
 
     /**
-     *
-     * @param bean
-     *            Spring Bean this object holds
-     * @param beanFactry
-     *            Spring factory that produced the bean with a ProxyReplacer.FACTORY_METHOD_NAME method
-     * @param paramClasses
-     * @param paramValues
+     * @param bean       Spring Bean this object holds
+     * @param beanFactory Spring factory that produced the bean with a ProxyReplacer.FACTORY_METHOD_NAME method
      */
-    public DetachableBeanHolder(Object bean, Object beanFactry, Class<?>[] paramClasses, Object[] paramValues) {
-        if (bean == null) {
-            LOGGER.error("Bean is null. The param value: {}", Arrays.toString(paramValues));
-        }
+    public DetachableBeanHolder(Object bean, Object beanFactory, Object beanName) {
```

Comment:
Object beanName should be 'String beanName', and beanName can never be null.

```diff
-    public DetachableBeanHolder(Object bean, Object beanFactory, Object beanName) {
+    public DetachableBeanHolder(Object bean, Object beanFactory, String beanName) {
         String beanNameStr = beanName.toString();
         this.bean = bean;
         this.beanFactory = beanFactory;
         this.beanName = beanNameStr;
 
-        if (beanName != null) {
-            synchronized (beanProxies) {
-                beanProxies.computeIfAbsent(beanNameStr, k -> new ArrayList<>());
-                beanProxies.get(beanName).add(new WeakReference<>(this));
-            }
+        synchronized (beanProxies) {
+            beanProxies.computeIfAbsent(beanNameStr, k -> new ArrayList<>());
+            beanProxies.get(beanName).add(new WeakReference<>(this));
         }
     }

```

Reply:
done

Reply:
done